Mathlib/CategoryTheory/Retract.lean

```diff
@@ -0,0 +1,142 @@
+abbrev RetractArrow {X Y Z W : C} (f : X ⟶ Y) (g : Z ⟶ W) := Retract (Arrow.mk f) (Arrow.mk g)
+
+namespace RetractArrow
+
+variable {X Y Z W : C} {f : X ⟶ Y} {g : Z ⟶ W} (h : RetractArrow f g)
+
+@[reassoc]
+lemma i_w : h.i.left ≫ g = f ≫ h.i.right := h.i.w
+
+@[reassoc]
+lemma r_w : h.r.left ≫ f = g ≫ h.r.right := h.r.w
+
+@[simp, reassoc]
+lemma retract_left : h.i.left ≫ h.r.left = 𝟙 X := Arrow.hom.congr_left h.retract
+
+@[simp, reassoc]
+lemma retract_right : h.i.right ≫ h.r.right = 𝟙 Y := Arrow.hom.congr_right h.retract
+
+@[reassoc]
+lemma fac : h.i.left ≫ g ≫ h.r.right = f := by simp
+
+/-- the bottom of a retract diagram determines a split epimorphism. -/
+@[simps] def splitEpi_left : SplitEpi h.r.left where
+  section_ := h.i.left
+
+/-- the top of a retract diagram determines a split epimorphism. -/
+@[simps] def splitEpi_right : SplitEpi h.r.right where
+  section_ := h.i.right
+
+/-- the bottom of a retract diagram determines a split monomorphism. -/
+@[simps] def splitMono_left : SplitMono h.i.left where
+  retraction := h.r.left
+
+/-- the top of a retract diagram determines a split monomorphism. -/
+@[simps] def splitMono_right : SplitMono h.i.right where
+  retraction := h.r.right
+
+instance : IsSplitEpi h.r.left := ⟨⟨h.splitEpi_left⟩⟩
+
+instance : IsSplitEpi h.r.right := ⟨⟨h.splitEpi_right⟩⟩
+
+instance : IsSplitMono h.i.left := ⟨⟨h.splitMono_left⟩⟩
+
+instance : IsSplitMono h.i.right := ⟨⟨h.splitMono_right⟩⟩
+
+end RetractArrow
+
+namespace MorphismProperty
+
+/-- A class of morphisms is stable under retracts if a retract of such a morphism still
+lies in the class. -/
+class IsStableUnderRetracts (P : MorphismProperty C) : Prop where
+  of_retract {X Y Z W : C} {f : X ⟶ Y} {g : Z ⟶ W} (h : RetractArrow f g) (hg : P g) : P f
+
+lemma of_retract {P : MorphismProperty C} [P.IsStableUnderRetracts]
+    {X Y Z W : C} {f : X ⟶ Y} {g : Z ⟶ W} (h : RetractArrow f g) (hg : P g) : P f :=
+  IsStableUnderRetracts.of_retract h hg
+
+instance IsStableUnderRetracts.monomorphisms : (monomorphisms C).IsStableUnderRetracts where
+  of_retract {_ _ _ _ f g} h (hg : Mono g) := ⟨fun α β w ↦ by
+    rw [← cancel_mono h.i.left, ← cancel_mono g, Category.assoc, Category.assoc,
+      h.i_w, reassoc_of% w]⟩
+
+instance IsStableUnderRetracts.epimorphisms : (epimorphisms C).IsStableUnderRetracts where
+  of_retract {_ _ _ _ f g} h (hg : Epi g) := ⟨fun α β w ↦ by
+    rw [← cancel_epi h.r.right, ← cancel_epi g, ← Category.assoc, ← Category.assoc, ← h.r_w,
+      Category.assoc, Category.assoc, w]⟩
+
+instance IsStableUnderRetracts.isomorphisms : (isomorphisms C).IsStableUnderRetracts where
+  of_retract {X Y Z W f g} h:= fun ⟨inv, ⟨h₁, h₂⟩⟩ ↦ ⟨by
+    refine ⟨h.i.right ≫ inv ≫ h.r.left, ?_, ?_⟩
+    · rw [← Category.assoc, ← h.i_w, Category.assoc, ← Category.assoc g, h₁,
+        Category.id_comp, h.retract_left]
+    · rw [Category.assoc, Category.assoc, h.r_w, ← Category.assoc inv, h₂, Category.id_comp,
+        h.retract_right]⟩
```

Comment:
In general, it is advisable not to use `Category.assoc` in the backwards direction, and here it is also better to use the `IsIso` API:
```suggestion
  of_retract {X Y Z W f g} h (_ : IsIso _) := by
    refine ⟨h.i.right ≫ inv g ≫ h.r.left, ?_, ?_⟩
    · rw [← h.i_w_assoc, IsIso.hom_inv_id_assoc, h.retract_left]
    · rw [Category.assoc, Category.assoc, h.r_w, IsIso.inv_hom_id_assoc, h.retract_right]
```